### ISSUE TITLE
This name is both non-const-lexical-bound and var-bound

### ISSUE THREAD

**Issue by Yoric** [binast/binjs-ref#335]
With the latest master, attempt to parse the following example:

```js
function f(t, e) {
}
if (foo) // NO braces here.
    function f(t, e) {
    }
```

This fails with

```
This name is both non-const-lexical-bound and var-bound: IdentifierName(Dynamic("f"))', crates/binjs_es6/src/scopes.rs:251:13
```

**Comment by Yoric** [comment]
@arai-a you have touched this code more recently than me, do you want to take a look?

**Comment by arai-a** [comment]
do you mean this is a regression?

**Comment by Yoric** [comment]
No, I mean that you may understand that piece of code better than I do these days :)

If I'm wrong, don't worry, I'll take a look at it this week.

**Comment by arai-a** [comment]
it's this case
https://tc39.github.io/ecma262/#sec-functiondeclarations-in-ifstatement-statement-clauses
and I think we don't handle Annex B things.

**Comment by Yoric** [comment]
See, I knew you knew that stuff better than me :)

Maybe we should find a way to fail gracefully instead of asserting?

**Comment by arai-a** [comment]
will try adding the handling, at least for strict mode.

**Comment by arai-a** [comment]
to properly handle non-strict-mode code, we need to add block statement, or something like that (maybe, new interface for implicit block?), in order to store scope information for the function.
that needs to be handled in the spec I think.
I'll address strict mode case here.

**Comment by RReverser** [comment]
> we need to add block statement

Could you elaborate please? Where would it be inserted?

In non-strict code the second function is meant to shadow the first one if the condition is true, so they should share the same scope.

**Comment by arai-a** [comment]
I misunderstood the current situation.
currently we're adding block with lexical scope for `if (foo) { function f() {} }`, but apparently that's wrong. (I was trying to align `if (foo) function f() {}` with that)

so, this should be fixed by changing the function declaration not-to-use lexical scope.
